src/core/batch.rkt

```diff
@@ -145,6 +153,86 @@
       [_ node]))
   (unmunge reg))
 
+(define (batch-restore-index batch)
+  (make-hash (for/list ([node (in-vector (batch-nodes batch))]
+                        [n (in-naturals)])
+               (cons node n))))
+
+(define (egg-nodes->batch egg-nodes id->spec input-batch rename-dict)
+  (define out (make-mutable-batch))
+  (set-mutable-batch-index! out (batch-restore-index input-batch))
+
+  ; This fuction here is only because of cycles in loads:( Can not be imported from egg-herbie.rkt
+  (define (egg-parsed->expr expr rename-dict type)
+    (let loop ([expr expr]
+               [type type])
+      (match expr
+        [(? number?) (if (representation? type) (literal expr (representation-name type)) expr)]
+        [(? symbol?)
+         (if (hash-has-key? rename-dict expr) (car (hash-ref rename-dict expr)) (list expr))]
+        [(list '$approx spec impl)
+         (define spec-type (if (representation? type) (representation-type type) type))
+         (approx (loop spec spec-type) (loop impl type))]
+        [(list 'if cond ift iff)
+         (if (representation? type)
+             (list 'if (loop cond (get-representation 'bool)) (loop ift type) (loop iff type))
+             (list 'if (loop cond 'bool) (loop ift type) (loop iff type)))]
+        [(list (? impl-exists? impl) args ...) (cons impl (map loop args (impl-info impl 'itype)))]
+        [(list op args ...) (cons op (map loop args (operator-info op 'itype)))])))
+
+  (define (eggref id)
+    (cdr (vector-ref egg-nodes id)))
+
+  (define (add-root node type)
+    (match node
+      [(? number?)
+       (batch-push! out (if (representation? type) (literal node (representation-name type)) node))]
+      [(? symbol?)
+       (batch-push! out (if (hash-has-key? rename-dict node) (car (hash-ref rename-dict node)) node))]
+      [(list '$approx spec impl)
+       (define spec* (vector-ref id->spec spec))
+       (unless spec*
+         (error 'regraph-extract-variants "no initial approx node in eclass"))
+       (define spec-type (if (representation? type) (representation-type type) type))
+       (define final-spec (egg-parsed->expr spec* rename-dict spec-type))
+       (batch-push! out (approx final-spec (add-root (eggref impl) type)))]
```

Comment:
This is the only "new" part. Can you talk me through why this differs from `egg-parsed->expr`?

Reply:
Yes, `egg-parsed->expr` recurse on `Expr`, while `add-root` recurses on nodes. + before calling `add-root` again, a new node needs to be extracted from `egg-nodes` (by `(eggref idx)`)

Reply:
I see, just juggling indices.